lib/ex_integrate/pipeline_runner.ex

```diff
@@ -2,14 +2,82 @@ defmodule ExIntegrate.Boundary.PipelineRunner do
   @moduledoc """
   Responsible for running steps and reporting their results.
   """
+  use GenServer, restart: :temporary
+  require Logger
 
-  alias ExIntegrate.Core.Step
+  @me __MODULE__
+  @task_supervisor ExIntegrate.TaskSupervisor
+
+  alias ExIntegrate.Boundary.StepRunner
   alias ExIntegrate.Core.Pipeline
 
-  @spec run_pipeline(Pipeline.t()) :: Pipeline.t()
+  # Client API
+
+  @spec start_link(Access.t()) :: {:ok, pid} | {:error, term} | :ignore
+  def start_link(opts) do
+    name = opts[:name] || @me
+    GenServer.start_link(__MODULE__, opts, name: name)
+  end
+
+  # GenServer callbacks
+
+  @impl GenServer
+  def init(opts) do
+    initial_state = Access.fetch!(opts, :pipeline) |> Pipeline.advance()
+
+    log = opts[:log] || true
+    config = %{log: log}
+
+    {:ok, {initial_state, config}, {:continue, {:run_step}}}
+  end
+
+  @impl GenServer
+  def handle_continue({:run_step}, {state, config}) do
+    Task.Supervisor.async_nolink(@task_supervisor, fn ->
+      current_step = Pipeline.current_step(state)
+      Logger.info("Starting step: #{inspect(current_step)}")
+      StepRunner.run_step(current_step, log: config.log)
+    end)
+
+    {:noreply, {state, config}}
+  end
+
+  @impl GenServer
+  def handle_info({ref, {:ok, step}}, {state, config}) do
+    Logger.info("Step completed. #{inspect(step)}")
+    Process.demonitor(ref, [:flush])
+
+    new_state =
+      state
+      |> Pipeline.replace_current_step(step)
+      |> Pipeline.advance()
+
+    if Pipeline.complete?(new_state) do
+      Logger.info("All steps completed successfully. Terminating pipeline #{inspect(new_state)}")
+      {:stop, :normal, {new_state, config}}
+    else
+      {:noreply, {new_state, config}, {:continue, {:run_step}}}
+    end
+  end
+
+  @impl GenServer
+  def handle_info({_ref, {:error, step}}, {state, config}) do
+    Logger.info("Step errored. #{inspect(step)}")
+    {:stop, :step_failure, state}
+  end
+
+  @impl GenServer
+  def handle_info({:DOWN, ref, _, _, reason}, {state, config}) do
+    Logger.info("Step task #{inspect(ref)} terminated unexpectedly. Reason: #{inspect(reason)}")
+    {:stop, :step_failure, state}
+  end
+
+  @deprecated """
```

Comment:
It seems like this could just be dropped so that we don't carry forward dead code.

Reply:
Agreed in principle. I guess my reason for deprecating was that the top-level `ExIntegrate`, and its corresponding tests, still utilize this sequential function to run a series of pipelines. If we removed this code now, we wouldn't be able to run our `ei.json` file anymore until we finished the rest of the OTP structure.

Stepping back, I guess we initially scaffolded a minimal, sequential implementation that could complete an entire run from an `ei.json` file, and now we are partway through the process of replacing that implementation with a concurrent one.

What remains is adding the higher-level modules: a RunManager for launching and responding to PipelineRunners, and a long-lived `ExIntegrate.Server` process that starts runs on demand by launching a RunManager.

If I understand correctly, once the OTP structure is in place, we can safely remove this code.

Reply:
@garrettmichaelgeorge makes sense now thanks!